### PR TITLE
[CARBONDATA-769] Added codegen support to spark2.1 carbon dictionary decoder.

### PR DESCRIPTION
**Pull request by ravipesala**

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1127/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 1]
```diff
@@ -200,6 +201,123 @@ case class CarbonDictionaryDecoder(
...
+                 """.stripMargin
+              case _ =>
+                s"""
+                   | UTF8String $value = UTF8String.fromBytes($valueIntern);
+                 """.stripMargin
+            }
+          code +=
+            s"""
+               |$caseCode
+             """.stripMargin
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: simplify it in one line

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 2]
```diff
@@ -200,6 +201,123 @@ case class CarbonDictionaryDecoder(
...
+      val cacheProvider: CacheProvider = CacheProvider.getInstance
+      val forwardDictionaryCache: Cache[DictionaryColumnUniqueIdentifier, Dictionary] =
+        cacheProvider.createCache(CacheType.FORWARD_DICTIONARY, storePath)
+      val dicts: Seq[ForwardDictionaryWrapper] = getDictionaryWrapper(absoluteTableIdentifiers,
+        forwardDictionaryCache, storePath)
+
+      val exprs = child.output.map(x =>
+        ExpressionCanonicalizer.execute(BindReferences.bindReference(x, child.output)))
+      ctx.currentVars = input
+      val resultVars = exprs.zipWithIndex.map { e =>
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: change `e` to `case (expr, index)`

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 3]
```diff
@@ -200,6 +201,123 @@ case class CarbonDictionaryDecoder(
...
+    }.toMap
+
+    if (isRequiredToDecode) {
+      val cacheProvider: CacheProvider = CacheProvider.getInstance
+      val forwardDictionaryCache: Cache[DictionaryColumnUniqueIdentifier, Dictionary] =
+        cacheProvider.createCache(CacheType.FORWARD_DICTIONARY, storePath)
+      val dicts: Seq[ForwardDictionaryWrapper] = getDictionaryWrapper(absoluteTableIdentifiers,
+        forwardDictionaryCache, storePath)
+
+      val exprs = child.output.map(x =>
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: change `map(x` to `map { x`

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 4]
```diff
@@ -225,6 +343,25 @@ case class CarbonDictionaryDecoder(
     dicts
   }
 
+  private def getDictionaryWrapper(atiMap: Map[String, AbsoluteTableIdentifier],
+      cache: Cache[DictionaryColumnUniqueIdentifier, Dictionary], storePath: String) = {
+    val dicts: Seq[ForwardDictionaryWrapper] = getDictionaryColumnIds.map { f =>
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: suggest to change `f` to a more meaningful variable

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala (hunk 3 above)
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala (hunk 4 above)
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1159/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala (hunk 2 above)
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1161/

**Pull request closed without merging**